google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -0,0 +1,386 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/pubsub/service"
+require "google/cloud/errors"
+require "monitor"
+require "forwardable"
+require "concurrent"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Subscriber
+      #
+      class Subscriber
+        include MonitorMixin
+
+        ##
+        # @private Implementation attributes.
+        attr_reader :request_queue, :output_enum, :thread_pool, :service
+
+        ##
+        # Subscriber attributes.
+        attr_reader :callback, :subscription_name, :deadline, :inventory,
+                    :threads
+
+        ##
+        # @private Create an empty {Subscriber} object.
+        def initialize callback, subscription_name, deadline, inventory,
+                       threads, service
+          @request_queue = nil
+          @output_enum = nil
+          @callback = callback
+          @subscription_name = subscription_name
+          @deadline = deadline || 60
+          @inventory = Inventory.new self, (inventory || 100)
+          @threads = threads || [2, Concurrent.processor_count * 2].max
+          @service = service
+
+          super() # to init MonitorMixin
+        end
+
+        def start
+          synchronize do
+            return if @request_queue
+
+            @request_queue = EnumeratorQueue.new self
+            @thread_pool = Concurrent::FixedThreadPool.new threads
+            start_streaming!
+          end
+
+          true
+        end
+
+        def stop
+          synchronize do
+            return if @request_queue.nil?
+            @request_queue.push self
+            @inventory.stop
+            @stopped = true
+          end
+
+          true
+        end
+
+        def stopped?
+          synchronize { @stopped }
+        end
+
+        def paused?
+          synchronize { @paused }
+        end
+
+        def wait!
+          synchronize do
+            @background_thread.join if @background_thread
+
+            @thread_pool.shutdown
+            @thread_pool.wait_for_termination
+          end
+
+          true
+        end
+
+        ##
+        # @private
+        def acknowledge *messages
```

Comment:
I would batch this so that messages are marked as 'acked' but then all sent together as one request. This isn't incorrect, but I think it would save CPU if you just had a background thread polling (could be 20x a second or something like that) instead of constructing a request for every message processed.